File: compilers/rust/src/frontend/sighash_directive.rs
//! `@sighash` directive parsing (issue #123).
//!
//! A public method may carry a `/** @sighash <FLAGS> */` comment directive that
//! declares which BIP-143 sighash type its auto-injected covenant (and the
//! SDK-built preimage) commits to. `<FLAGS>` is a `|`-separated set of SigHash
//! names, e.g. `SINGLE|FORKID`, `ALL|ANYONECANPAY|FORKID`, `NONE|FORKID`.
//!
//! The default (no directive) is `ALL|FORKID` (0x41) — byte-identical to the
//! historically-pinned mode, so existing fixtures see ZERO change.
//!
//! This module owns the *flag grammar* (name → value, combo validity); the
//! *detection* lives in the parser, mirroring the TypeScript reference module
//! `packages/runar-compiler/src/passes/sighash-directive.ts`.

/// Numeric value of each sighash flag name.
fn flag_value(name: &str) -> Option<i64> {
    match name {
        "ALL" => Some(0x01),
        "NONE" => Some(0x02),
        "SINGLE" => Some(0x03),
        "FORKID" => Some(0x40),
        "ANYONECANPAY" => Some(0x80),
        _ => None,
    }
}

/// True if `name` is a base-type name. Exactly one MUST appear in a directive.
fn is_base_type_name(name: &str) -> bool {
    matches!(name, "ALL" | "NONE" | "SINGLE")
}

/// SIGHASH_ALL | SIGHASH_FORKID — the default when no directive is present.
pub const SIGHASH_DEFAULT: i64 = 0x41;

/// Base-type mask. `value & BASE_TYPE_MASK` recovers 1/2/3 (ALL/NONE/SINGLE)
/// after the FORKID/ANYONECANPAY high bits are stripped.
pub const BASE_TYPE_MASK: i64 = 0x1f;
pub const BASE_ALL: i64 = 0x01;
pub const BASE_NONE: i64 = 0x02;
pub const BASE_SINGLE: i64 = 0x03;
pub const FLAG_FORKID: i64 = 0x40;
pub const FLAG_ANYONECANPAY: i64 = 0x80;

/// Parse the flag list of an `@sighash` directive.
///
/// `flags_text` is the raw text following `@sighash` (e.g. `"SINGLE|FORKID"`),
/// with any trailing comment punctuation already stripped by the caller.
///
/// Validation (security-relevant — a mis-declared mode is an exploit class):
///   - every name must be a known flag (reject typos like `FORKD`)
///   - EXACTLY ONE base type (ALL/NONE/SINGLE) — reject zero, and reject
///     nonsensical combos such as `ALL|NONE`. Checked on NAMES, not the OR-ed
///     numeric value, because `ALL|NONE` (0x01|0x02) collides with the numeric
///     value of SINGLE (0x03) — a silent, dangerous aliasing.
///   - reject a duplicated flag name (signals a copy/paste error).
///   - FORKID is mandatory on BSV: a FORKID-less flag set deploys to brick.
pub fn parse_sighash_flags(flags_text: &str) -> Result<i64, String> {
    let raw = flags_text.trim();
    if raw.is_empty() {
        return Err(
            "@sighash directive requires at least one flag (e.g. `@sighash ALL|FORKID`)".to_string(),
        );
    }

    let names: Vec<&str> = raw.split('|').map(|n| n.trim()).collect();
    let mut seen: Vec<&str> = Vec::new();
    let mut base_types: Vec<&str> = Vec::new();
    let mut value: i64 = 0;

    for name in &names {
        if name.is_empty() {
            return Err(format!("@sighash directive has an empty flag in \"{}\"", raw));
        }
        let v = match flag_value(name) {
            Some(v) => v,
            None => {
                return Err(format!(
                    "@sighash: unknown flag \"{}\" (valid: ALL, NONE, SINGLE, FORKID, ANYONECANPAY)",
                    name
                ));
            }
        };
        if seen.contains(name) {
            return Err(format!("@sighash: duplicate flag \"{}\" in \"{}\"", name, raw));
        }
        seen.push(name);
        if is_base_type_name(name) {
            base_types.push(name);
        }
        value |= v;
    }

    if base_types.is_empty() {
        return Err(format!(
            "@sighash: must specify exactly one base type (ALL, NONE, or SINGLE); got \"{}\"",
            raw
        ));
    }
    if base_types.len() > 1 {
        return Err(format!(
            "@sighash: cannot combine base types ({}) — pick exactly one of ALL/NONE/SINGLE",
            base_types.join("|")
        ));
    }

    // FORKID is mandatory on BSV: the entire OP_PUSH_TX / BIP-143 preimage
    // machinery is FORKID-only, so a FORKID-less flag set deploys a covenant
    // whose derived signature can never verify (deploy-to-brick).
    if (value & FLAG_FORKID) == 0 {
        return Err(format!(
            "@sighash: FORKID is mandatory on BSV; write e.g. @sighash {}|FORKID (got \"{}\")",
            base_types[0], raw
        ));
    }

    Ok(value)
}

/// Extract and parse an `@sighash` directive from a block of comment text.
/// Returns `None` when no well-formed `@sighash <FLAGS>` token is present,
/// otherwise the parse result (value or error). Mirrors the TS reference regex
/// `/@sighash\s+([A-Za-z0-9_|\s]*?)(?:\*\/|\n|\r|$)/`.
pub fn extract_sighash_directive(comment_text: &str) -> Option<Result<i64, String>> {
    let idx = comment_text.find("@sighash")?;
    let after = &comment_text[idx + "@sighash".len()..];

    // Segment = text up to the first terminator (`*/`, newline, CR, or end).
    let mut end = after.len();
    for (i, ch) in after.char_indices() {
        if ch == '\n' || ch == '\r' {
            end = i;
            break;
        }
        if ch == '*' && after[i + 1..].starts_with('/') {
            end = i;
            break;
        }
    }
    let seg = &after[..end];

    // The reference regex requires `\s+` immediately after `@sighash`, and the
    // captured flag list may only contain `[A-Za-z0-9_|\s]`. If the leading
    // whitespace is missing or the segment carries any other character, the
    // regex fails to match → no directive (falls back to the default).
    if !seg.starts_with(|c: char| c == ' ' || c == '\t' || c == '\n' || c == '\r') {
        return None;
    }
    if !seg
        .chars()
        .all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '|' || c.is_whitespace())
    {
        return None;
    }

    Some(parse_sighash_flags(seg.trim()))
}

/// Human-readable rendering of a sighash value (for diagnostics).
pub fn describe_sighash(value: i64) -> String {
    let mut parts: Vec<String> = Vec::new();
    let base = value & BASE_TYPE_MASK;
    if base == BASE_ALL {
        parts.push("ALL".to_string());
    } else if base == BASE_NONE {
        parts.push("NONE".to_string());
    } else if base == BASE_SINGLE {
        parts.push("SINGLE".to_string());
    } else {
        parts.push(format!("0x{:x}", base));
    }
    if value & FLAG_ANYONECANPAY != 0 {
        parts.push("ANYONECANPAY".to_string());
    }
    if value & FLAG_FORKID != 0 {
        parts.push("FORKID".to_string());
    }
    parts.join("|")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_single_forkid() {
        assert_eq!(parse_sighash_flags("SINGLE|FORKID"), Ok(0x43));
    }

    #[test]
    fn parses_all_anyonecanpay_forkid() {
        assert_eq!(parse_sighash_flags("ALL|ANYONECANPAY|FORKID"), Ok(0xC1));
    }

    #[test]
    fn parses_default_all_forkid() {
        assert_eq!(parse_sighash_flags("ALL|FORKID"), Ok(SIGHASH_DEFAULT));
    }

    #[test]
    fn rejects_unknown_flag() {
        assert!(parse_sighash_flags("FORKD|ALL").unwrap_err().contains("unknown flag"));
    }

    #[test]
    fn rejects_all_none_alias() {
        // ALL|NONE numerically collides with SINGLE (0x03); caught on names.
        assert!(parse_sighash_flags("ALL|NONE|FORKID")
            .unwrap_err()
            .contains("cannot combine base types"));
    }

    #[test]
    fn rejects_no_base_type() {
        assert!(parse_sighash_flags("FORKID")
            .unwrap_err()
            .contains("must specify exactly one base type"));
    }

    #[test]
    fn rejects_duplicate_flag() {
        assert!(parse_sighash_flags("ALL|ALL|FORKID")
            .unwrap_err()
            .contains("duplicate flag"));
    }

    #[test]
    fn rejects_missing_forkid() {
        assert!(parse_sighash_flags("SINGLE").unwrap_err().contains("FORKID is mandatory"));
    }

    #[test]
    fn extracts_from_jsdoc_block() {
        assert_eq!(
            extract_sighash_directive("* @sighash SINGLE|FORKID "),
            Some(Ok(0x43))
        );
    }

    #[test]
    fn extract_none_when_absent() {
        assert_eq!(extract_sighash_directive("* just a comment"), None);
    }

    #[test]
    fn describe_roundtrip() {
        assert_eq!(describe_sighash(0x43), "SINGLE|FORKID");
        assert_eq!(describe_sighash(0xC1), "ALL|ANYONECANPAY|FORKID");
    }
}

File: compilers/java/src/main/java/runar/compiler/passes/SighashDirective.java
package runar.compiler.passes;

import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * {@code @sighash} directive parsing (issue #123) — Java port of
 * {@code packages/runar-compiler/src/passes/sighash-directive.ts}.
 *
 * <p>A public method may carry a {@code /** @sighash <FLAGS> *&#47;} comment
 * directive that declares which BIP-143 sighash type its auto-injected covenant
 * (and the SDK-built preimage) commits to. {@code <FLAGS>} is a {@code |}-separated
 * set of SigHash names, e.g. {@code SINGLE|FORKID}, {@code ALL|ANYONECANPAY|FORKID},
 * {@code NONE|FORKID}.
 *
 * <p>The default (no directive) is {@code ALL|FORKID} (0x41) — byte-identical to
 * the historically-pinned mode, so existing fixtures see ZERO change. Detection
 * lives in the parser; this class owns the flag grammar (name -&gt; value, combo
 * validity, FORKID-mandatory).
 */
public final class SighashDirective {
    private SighashDirective() {}

    /** Numeric value of each sighash flag name. */
    private static final Map<String, Integer> FLAG_VALUES = new LinkedHashMap<>();
    static {
        FLAG_VALUES.put("ALL", 0x01);
        FLAG_VALUES.put("NONE", 0x02);
        FLAG_VALUES.put("SINGLE", 0x03);
        FLAG_VALUES.put("FORKID", 0x40);
        FLAG_VALUES.put("ANYONECANPAY", 0x80);
    }

    /** The base-type names. Exactly one MUST appear in a directive. */
    private static final Set<String> BASE_TYPE_NAMES = Set.of("ALL", "NONE", "SINGLE");

    /** SIGHASH_ALL | SIGHASH_FORKID — the default when no directive is present. */
    public static final int SIGHASH_DEFAULT = 0x41;

    /**
     * Base-type mask. {@code sigHashType & BASE_TYPE_MASK} recovers 1/2/3
     * (ALL/NONE/SINGLE) after the FORKID/ANYONECANPAY high bits are stripped.
     */
    public static final int BASE_TYPE_MASK = 0x1f;
    public static final int BASE_ALL = 0x01;
    public static final int BASE_NONE = 0x02;
    public static final int BASE_SINGLE = 0x03;
    public static final int FLAG_FORKID = 0x40;
    public static final int FLAG_ANYONECANPAY = 0x80;

    /** Parse result: exactly one of {@code value} / {@code error} is non-null. */
    public record Result(Integer value, String error) {
        static Result ok(int v) { return new Result(v, null); }
        static Result err(String e) { return new Result(null, e); }
        public boolean isError() { return error != null; }
    }

    /**
     * Parse the flag list of an {@code @sighash} directive.
     *
     * <p>{@code flagsText} is the raw text following {@code @sighash} (e.g.
     * {@code "SINGLE|FORKID"}). Validation (security-relevant — a mis-declared
     * mode is an exploit class):
     * <ul>
     *   <li>every name must be a known flag (reject typos like {@code FORKD})</li>
     *   <li>EXACTLY ONE base type (ALL/NONE/SINGLE) — reject zero, and reject
     *       nonsensical combos such as {@code ALL|NONE}. Checked on NAMES, not
     *       the OR-ed numeric value, because {@code ALL|NONE} (0x01|0x02)
     *       collides with the numeric value of SINGLE (0x03).</li>
     *   <li>reject a duplicated flag name.</li>
     *   <li>FORKID is mandatory on BSV (deploy-to-brick otherwise).</li>
     * </ul>
     */
    public static Result parseSighashFlags(String flagsText) {
        String raw = flagsText == null ? "" : flagsText.trim();
        if (raw.isEmpty()) {
            return Result.err("@sighash directive requires at least one flag (e.g. `@sighash ALL|FORKID`)");
        }

        String[] names = raw.split("\\|", -1);
        Set<String> seen = new LinkedHashSet<>();
        java.util.List<String> baseTypes = new java.util.ArrayList<>();
        int value = 0;

        for (String rawName : names) {
            String name = rawName.trim();
            if (name.isEmpty()) {
                return Result.err("@sighash directive has an empty flag in \"" + raw + "\"");
            }
            if (!FLAG_VALUES.containsKey(name)) {
                return Result.err(
                    "@sighash: unknown flag \"" + name
                    + "\" (valid: ALL, NONE, SINGLE, FORKID, ANYONECANPAY)");
            }
            if (seen.contains(name)) {
                return Result.err("@sighash: duplicate flag \"" + name + "\" in \"" + raw + "\"");
            }
            seen.add(name);
            if (BASE_TYPE_NAMES.contains(name)) {
                baseTypes.add(name);
            }
            value |= FLAG_VALUES.get(name);
        }

        if (baseTypes.isEmpty()) {
            return Result.err(
                "@sighash: must specify exactly one base type (ALL, NONE, or SINGLE); got \"" + raw + "\"");
        }
        if (baseTypes.size() > 1) {
            return Result.err(
                "@sighash: cannot combine base types (" + String.join("|", baseTypes)
                + ") — pick exactly one of ALL/NONE/SINGLE");
        }

        // FORKID is mandatory on BSV: the entire OP_PUSH_TX / BIP-143 preimage
        // machinery is FORKID-only, so a FORKID-less flag set deploys a covenant
        // whose derived signature can never verify (deploy-to-brick).
        if ((value & FLAG_VALUES.get("FORKID")) == 0) {
            return Result.err(
                "@sighash: FORKID is mandatory on BSV; write e.g. @sighash "
                + baseTypes.get(0) + "|FORKID (got \"" + raw + "\")");
        }

        return Result.ok(value);
    }

    private static final Pattern SIGHASH_RE =
        Pattern.compile("@sighash\\s+([A-Za-z0-9_|\\s]*?)(?:\\*/|\\n|\\r|$)");

    /**
     * Extract and parse an {@code @sighash} directive from a block of comment
     * text. Returns {@code null} when no {@code @sighash} token is present,
     * otherwise the parse result (value or error).
     */
    public static Result extractSighashDirective(String commentText) {
        if (commentText == null) {
            return null;
        }
        Matcher m = SIGHASH_RE.matcher(commentText);
        if (!m.find()) {
            return null;
        }
        return parseSighashFlags(m.group(1) == null ? "" : m.group(1));
    }

    /** Human-readable rendering of a sighash value (for diagnostics). */
    public static String describeSighash(int value) {
        java.util.List<String> parts = new java.util.ArrayList<>();
        int base = value & BASE_TYPE_MASK;
        if (base == BASE_ALL) {
            parts.add("ALL");
        } else if (base == BASE_NONE) {
            parts.add("NONE");
        } else if (base == BASE_SINGLE) {
            parts.add("SINGLE");
        } else {
            parts.add("0x" + Integer.toHexString(base));
        }
        if ((value & FLAG_ANYONECANPAY) != 0) {
            parts.add("ANYONECANPAY");
        }
        if ((value & FLAG_FORKID) != 0) {
            parts.add("FORKID");
        }
        return String.join("|", parts);
    }
}
